iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp

```diff
@@ -0,0 +1,317 @@
+static VectorType getVecType(OpBuilder &builder, Type t) {
+  if (!t.isa<IntegerType, FloatType>()) return {};
+  if (t.getIntOrFloatBitWidth() != 32) return {};
+  Type newElemType = t.isa<IntegerType>() ? builder.getI32Type().cast<Type>()
+                                          : builder.getF32Type().cast<Type>();
+  return VectorType::get(kVecSize, newElemType);
+}
+
+static MemRefType getVectorizedMemRefType(OpBuilder &builder, MemRefType type) {
+  Type elemType = type.getElementType();
+  VectorType vecType = getVecType(builder, elemType);
+  if (!vecType) return {};
+  unsigned elemSize = elemType.getIntOrFloatBitWidth();
+  unsigned vecSize = kVectorizationSizeInBits / elemSize;
+  SmallVector<int64_t, 2> newShape(type.getShape().begin(),
+                                   type.getShape().end());
+  if (newShape.empty()) return {};
+  if (newShape.back() % vecSize != 0) return {};
+  newShape.back() = newShape.back() / vecSize;
+  return MemRefType::get(newShape, vecType, {}, type.getMemorySpace());
+}
+
+Value legalizeToVectorType(OpBuilder &builder, Value val) {
+  Type type = val.getType();
+  if (type.isa<VectorType>()) {
+    return val;
+  } else if (type.isIntOrFloat()) {
+    auto vecType = getVecType(builder, type);
+    if (!vecType) return {};
+    // TODO(hanchung): Add a folder on vector::BroadcastOp so we don't need to
+    // create manually.
+    if (auto cst = val.getDefiningOp<ConstantOp>()) {
+      auto cstVecValue = DenseElementsAttr::get(vecType, cst.value());
+      return builder.create<ConstantOp>(val.getLoc(), vecType, cstVecValue)
+          .getResult();
+    }
+    return builder.create<vector::BroadcastOp>(val.getLoc(), vecType, val)
+        .getResult();
+  }
+  return {};
+}
+
+template <typename OpTy>
+struct VectorizeElementwiseOp : public OpConversionPattern<OpTy> {
+  using OpConversionPattern<OpTy>::OpConversionPattern;
+  LogicalResult matchAndRewrite(
+      OpTy op, ArrayRef<Value> args,
+      ConversionPatternRewriter &rewriter) const override {
+    if (llvm::all_of(args, [](Value arg) {
+          return arg.getType().isIntOrIndexOrFloat();
+        })) {
+      return failure();
+    }
+    SmallVector<Value, 4> vecArgs;
+    for (Value arg : args) {
+      Value val = legalizeToVectorType(rewriter, arg);
+      if (!val) return failure();
+      vecArgs.push_back(val);
+    }
+    auto vecType = getVecType(rewriter, op.getResult().getType());
+    if (!vecType) return failure();
+    auto newOp = rewriter.create<OpTy>(op.getLoc(), vecType, vecArgs);
+    rewriter.replaceOp(op, newOp.getOperation()->getResults());
+    return success();
+  }
+};
+
+template <typename OpTy>
+struct VectorizeCmpOp : public OpConversionPattern<OpTy> {
+  using OpConversionPattern<OpTy>::OpConversionPattern;
+  LogicalResult matchAndRewrite(
+      OpTy op, ArrayRef<Value> args,
+      ConversionPatternRewriter &rewriter) const override {
+    if (llvm::all_of(args, [](Value arg) {
```

Comment:
This is done in every class. Maybe have a base `VectorizeOpBase` that checks this and calls an apply method to do the vectorization.

It would have been really nice to use [this](https://stackoverflow.com/questions/50754525/pass-an-array-variable-as-template-argument) trick to have a single implementation for all these classes, and just set the arguments to vectorize during instantiation. Seems like a C++-17 feature though.

Reply:
Sounds good, thanks!